src/python/pants/backend/jvm/subsystems/zinc.py

```diff
@@ -153,6 +221,102 @@ def compiler_interface(self):
+    """Return the path to the scala library jar (runtime).
+
+    :rtype: str
+    """
+    return self._zinc_factory._scala_reflect(self._products)
+
+  @memoized_property
+  def _compiler_bridge_cache_dir(self):
+    """A directory where we can store compiled copies of the `compiler-bridge`.
+
+    The compiler-bridge is specific to each scala version.
+    Currently we compile the `compiler-bridge` only once, while bootstrapping.
+    Then, we store it in the working directory under .pants.d/zinc/<cachekey>, where
+    <cachekey> is calculated using the locations of zinc, the compiler interface,
+    and the compiler bridge.
+    """
+    hasher = sha1()
+    for cp_entry in [self.zinc, self.compiler_interface, self.compiler_bridge]:
+      hasher.update(os.path.relpath(cp_entry, self.workdir))
+    key = hasher.hexdigest()[:12]
+
+    return os.path.join(self.workdir, 'zinc', key)
+
+  def _make_relative(self, path):
+    """A utility function to create relative paths to the work dir"""
+    return fast_relpath(path, self.workdir)
+
+  @memoized_property
+  def workdir(self):
+    """A utility method to access the current workdir"""
+    return self._zinc_factory.get_options().pants_workdir
+
+  def compile_compiler_bridge(self, context):
+    """Compile the compiler bridge to be used by zinc, using our scala bootstrapper.
+    It will compile and cache the jar, and materialize it if not already there.
+
+    :param context: The context of the task trying to compile the bridge.
+                    This is mostly needed to use its scheduler to create digests of the relevant jars.
+    :return: The absolute path to the compiled scala-compiler-bridge jar.
+    """
+    bridge_jar = os.path.join(self._compiler_bridge_cache_dir, 'scala-compiler-bridge.jar')
+    bootstrapper = self._zinc_factory._compiler_bootstrapper(self._products)
+
+    bootstrapper_args = [
+      bridge_jar,
+      self.compiler_interface,
+      self.compiler_bridge,
+      self.scala_compiler,
+      self.scala_library,
+      self.scala_reflect,
+    ]
+    input_jar_snapshots = context._scheduler.capture_snapshots((PathGlobsAndRoot(
+      PathGlobs(tuple([self._make_relative(jar) for jar in bootstrapper_args])),
+      text_type(self.workdir)
+    ),))
+    input_jars_digest = context._scheduler.merge_directories(
+      tuple(s.directory_digest for s in (input_jar_snapshots))
+    )
+    argv = tuple(['.jdk/bin/java'] +
+                 ['-cp', bootstrapper, Zinc.ZINC_BOOTSTRAPER_MAIN] +
+                 bootstrapper_args
+    )
+    req = ExecuteProcessRequest(
+      argv=argv,
+      input_files=input_jars_digest,
+      output_files=(self._make_relative(bridge_jar),),
+      output_directories=(self._make_relative(self._compiler_bridge_cache_dir),),
+      description="bootstrap compiler bridge.",
+      jdk_home=text_type(self.dist.home),
+    )
+    res = context.execute_process_synchronously(req, 'zinc-subsystem', [WorkUnitLabel.COMPILER])
```

Comment:
I think I lean against making `execute_process_synchronously` raise an exception, as then it would be doing judgement calls about the value of the return code. I think the best thing to address this is to create a helper function to call the process and raise if it fails, or something of the sort.